### ISSUE TITLE
Server very slow for certain repositories

### ISSUE THREAD

**Issue by tripodsan** [adobe/git-server#233]
**Description**
the bigger a repository gets, the slower the server, even for files that are on disk.

example repository:
```
$ git-sizer --verbose
Processing blobs: 808
Processing trees: 1368
Processing commits: 483
Matching commits to trees: 483
Processing annotated tags: 0
Processing references: 15
| Name                         | Value     | Level of concern               |
| ---------------------------- | --------- | ------------------------------ |
| Overall repository size      |           |                                |
| * Commits                    |           |                                |
|   * Count                    |   483     |                                |
|   * Total size               |   194 KiB |                                |
| * Trees                      |           |                                |
|   * Count                    |  1.37 k   |                                |
|   * Total size               |   507 KiB |                                |
|   * Total tree entries       |  13.4 k   |                                |
| * Blobs                      |           |                                |
|   * Count                    |   808     |                                |
|   * Total size               |   124 MiB |                                |
| * Annotated tags             |           |                                |
|   * Count                    |     0     |                                |
| * References                 |           |                                |
|   * Count                    |    15     |                                |
|                              |           |                                |
| Biggest objects              |           |                                |
| * Commits                    |           |                                |
|   * Maximum size         [1] |  1.18 KiB |                                |
|   * Maximum parents      [2] |     2     |                                |
| * Trees                      |           |                                |
|   * Maximum entries      [3] |    53     |                                |
| * Blobs                      |           |                                |
|   * Maximum size         [4] |  13.5 MiB | *                              |
|                              |           |                                |
| History structure            |           |                                |
| * Maximum history depth      |   301     |                                |
| * Maximum tag depth          |     0     |                                |
|                              |           |                                |
| Biggest checkouts            |           |                                |
| * Number of directories  [5] |    58     |                                |
| * Maximum path depth     [6] |     5     |                                |
| * Maximum path length    [7] |    64 B   |                                |
| * Number of files        [8] |   231     |                                |
| * Total size of files    [9] |   104 MiB |                                |
| * Number of symlinks         |     0     |                                |
| * Number of submodules       |     0     |                                |
```

- fetching a file that exists: about 1s
- fetching a file that does not exist: about 300ms

----

another repository that is not that big:

```
$ git-sizer --verbose
Processing blobs: 1283
Processing trees: 1386
Processing commits: 963
Matching commits to trees: 963
Processing annotated tags: 0
Processing references: 34
| Name                         | Value     | Level of concern               |
| ---------------------------- | --------- | ------------------------------ |
| Overall repository size      |           |                                |
| * Commits                    |           |                                |
|   * Count                    |   963     |                                |
|   * Total size               |   372 KiB |                                |
| * Trees                      |           |                                |
|   * Count                    |  1.39 k   |                                |
|   * Total size               |  1.08 MiB |                                |
|   * Total tree entries       |  28.3 k   |                                |
| * Blobs                      |           |                                |
|   * Count                    |  1.28 k   |                                |
|   * Total size               |  41.2 MiB |                                |
| * Annotated tags             |           |                                |
|   * Count                    |     0     |                                |
| * References                 |           |                                |
|   * Count                    |    34     |                                |
|                              |           |                                |
| Biggest objects              |           |                                |
| * Commits                    |           |                                |
|   * Maximum size         [1] |  1.50 KiB |                                |
|   * Maximum parents      [2] |     2     |                                |
| * Trees                      |           |                                |
|   * Maximum entries      [3] |    57     |                                |
| * Blobs                      |           |                                |
|   * Maximum size         [4] |  1.22 MiB |                                |
|                              |           |                                |
| History structure            |           |                                |
| * Maximum history depth      |   638     |                                |
| * Maximum tag depth          |     0     |                                |
|                              |           |                                |
| Biggest checkouts            |           |                                |
| * Number of directories  [5] |    22     |                                |
| * Maximum path depth     [6] |     7     |                                |
| * Maximum path length    [5] |   164 B   | *                              |
| * Number of files        [7] |   120     |                                |
| * Total size of files    [8] |  5.72 MiB |                                |
| * Number of symlinks         |     0     |                                |
| * Number of submodules       |     0     |                                |
```

- fetching non existent file: 20ms
- fetching existent file: 30ms

**Comment by stefan-guggisberg** [comment]
Can you provide a test case?

**Comment by tripodsan** [comment]
> Can you provide a test case?

not really. I can't attach the 120mb repo :-) ? but you can try it with: https://github.com/davidnuescheler/pages
just do a `hlx up` and then you see the git-server started and you can try to time the requests.

**Comment by tripodsan** [comment]
@stefan-guggisberg is there some sort of disk-cache? I tried to debug the server, but now it's always fast :-)

**Comment by tripodsan** [comment]
I timed it and it seems that `git.status()` got really slow:

#### running ab -n 10 -c 1 

```
status helix-markup.yaml 505
resolve helix-markup.yaml 1
status helix-markup.yaml 314
resolve helix-markup.yaml 1
status helix-markup.yaml 335
resolve helix-markup.yaml 0
status helix-markup.yaml 359
resolve helix-markup.yaml 1
status helix-markup.yaml 347
resolve helix-markup.yaml 0
status helix-markup.yaml 394
resolve helix-markup.yaml 1
status helix-markup.yaml 339
resolve helix-markup.yaml 1
status helix-markup.yaml 388
resolve helix-markup.yaml 1
status helix-markup.yaml 353
resolve helix-markup.yaml 1
status helix-markup.yaml 336
```

#### running ab -n 10 -c 4

resolve helix-markup.yaml 1
status helix-markup.yaml 471
resolve helix-markup.yaml 2
resolve helix-markup.yaml 1
resolve helix-markup.yaml 1
resolve helix-markup.yaml 1
status helix-markup.yaml 1422
status helix-markup.yaml 1427
status helix-markup.yaml 1431
status helix-markup.yaml 1436
resolve helix-markup.yaml 1
resolve helix-markup.yaml 2
resolve helix-markup.yaml 2
resolve helix-markup.yaml 1
status helix-markup.yaml 1190
status helix-markup.yaml 1341
status helix-markup.yaml 1341
status helix-markup.yaml 1344
resolve helix-markup.yaml 0
status helix-markup.yaml 315

**Comment by tripodsan** [comment]
further timing iso-git:

```
resolve helix-markup.yaml 0
>>> status
>> ignore 1
>> getHeadTree 424
>> getOidAtPath 0
>> GitIndexManager.acquire 12
>> lstat 0
status helix-markup.yaml 437
```

**Comment by stefan-guggisberg** [comment]
>  I can't attach the 120mb repo :-)

but you can provide a url ;-)

**Comment by stefan-guggisberg** [comment]
> further timing iso-git:

can you provide the test source code?

**Comment by tripodsan** [comment]
> > I can't attach the 120mb repo :-)
> 
> but you can provide a url ;-)

see above: https://github.com/davidnuescheler/pages

**Comment by tripodsan** [comment]
> > further timing iso-git:
> 
> can you provide the test source code?

```
curl http://127.0.0.1:62716/raw/helix/github.com--davidnuescheler--pages/master/helix-markup.yaml
```

**Comment by stefan-guggisberg** [comment]
thank you, very helpful, but not exactly what I meant.

**Comment by tripodsan** [comment]
well, I added the timing code directly to the source code of isomophic-git in the node_modules folder,
and then I did just request the file. after I did a `npm install`, my timing code was gone. but I added it in the `index.cjs` in the `status()` function.

**Comment by stefan-guggisberg** [comment]
> I timed it and it seems that git.status() got really slow, and even worse if invoked concurrently:

The 'big' repository in question has a 126mb pack file. Slow concurrent operations could be related to  https://github.com/isomorphic-git/isomorphic-git/issues/1241#issuecomment-707057262.

**Comment by stefan-guggisberg** [comment]
[isomorphic-git v1.7.5](https://github.com/isomorphic-git/isomorphic-git/releases/tag/v1.7.5) introduced a regression which causes a massive performance degradation in `git-server`:
- requests to non-existing files (`404`) are now ~20-25x slower, i.e. ~220ms vs ~10ms 
- requests to existing files (`200`) are now ~70x slower, i.e. ~700ms vs ~10ms

**Comment by stefan-guggisberg** [comment]
short term fix: use version `1.7.4` in `git-server` and `helix-cli` until the underlying issue is fixed in `isomorphic-git`.

**Comment by stefan-guggisberg** [comment]
underlying `isomorphic-git` issue: https://github.com/isomorphic-git/isomorphic-git/issues/1251
